### DISCUSSION TITLE
Dynamically assert exception type with xUnit Theory

### DISCUSSION THREAD

**Discussion by sgabler** [fluentassertions/fluentassertions#2049] [Q&A] [answered]
I'm trying to find an elegant solution for my current problem, but I'm probably approaching this incorrectly. Maybe one of you has a good idea how to improve this?

Here's my code:

```csharp
public class OAuthServiceTests
{
    [Theory]
    [MemberData(nameof(InvalidProcessAccessTokenInput))]
    public async Task ProcessAccessTokenInput_should_throw_exception_for_invalid_input(TokenController.AccessTokenInput input, Type expectedException)
    {
        // Arrange
        var oAuthService = new OAuthService(null);
        
        // Act
        Func<Task> act = () => oAuthService.ProcessAccessTokenInput(input, string.Empty, CancellationToken.None);

        // Assert
        await act.Should().ThrowAsync<expectedException>();
    }

    public static TheoryData<TokenController.AccessTokenInput, Type> InvalidProcessAccessTokenInput =>
        new()
        {
            { new TokenController.AccessTokenInput { GrantType = "wrong" }, typeof(UnsupportedGrantTypeException) },
            { new TokenController.AccessTokenInput { }, typeof(InvalidInputException) }
        };
}
```

Of course this doesn't compile. I didn't find a method in FluentAssertions that I could use instead.

![image](https://user-images.githubusercontent.com/3625477/204826913-3053783d-1181-4fdb-82cc-a8daefeb5223.png)

But I hope from the example it's clear what I want to achieve. In short: I want to pass the expected exception type in the Theory, and then do the assertion with FluentAssertions.

I searched, but it seems that this is not expected/recommended behavior. Do I have to write a switch statement for each exception type instead?

**Comment by jnyrup** [comment, marked as answer]
You're so close!

In our test suite we use have this test using a generic test and non-generic memberdata.

https://github.com/fluentassertions/fluentassertions/blob/17c52d0e2d6adb366fbf8b57eabe69d7c3d2b131/Tests/FluentAssertions.Specs/Exceptions/ExceptionAssertionSpecs.cs#L31-L38

I just checked and you should be able to use

```csharp
[Theory]
[MemberData(nameof(InvalidProcessAccessTokenInput))]
public async Task ProcessAccessTokenInput_should_throw_exception_for_invalid_input<T>(TokenController.AccessTokenInput input, T _)
    where T : Exception
{
    // Arrange
    var oAuthService = new OAuthService(null);
    
    // Act
    Func<Task> act = () => oAuthService.ProcessAccessTokenInput(input, string.Empty, CancellationToken.None);

    // Assert
    await act.Should().ThrowAsync<T>();
}

public static TheoryData<TokenController.AccessTokenInput, Exception> InvalidProcessAccessTokenInput => new()
{
    { new TokenController.AccessTokenInput { GrantType = "wrong" }, new UnsupportedGrantTypeException() },
    { new TokenController.AccessTokenInput { }, new InvalidInputException() }
};
```

**Reply by sgabler** [reply]
Thanks, that works perfectly! No ugly switch statement anymore 👍
